providers/vercel/vercelProvider.go

```diff
@@ -0,0 +1,397 @@
+				// thus we are most likely can go through both.
+				req, err := toVercelCreateRequest(domain, newRec)
+				if err != nil {
+					return err
+				}
+				_, err = c.CreateDNSRecord(ctx, req)
+				return err
+			}
+
+			req, err := toVercelUpdateRequest(newRec)
+			if err != nil {
+				return err
+			}
+			_, err = c.UpdateDNSRecord(ctx, existingID, req)
+			return err
+		},
+	}
+}
+
+func (c *vercelProvider) mkDeleteCorrection(domain string, oldRec *models.RecordConfig, msg string) *models.Correction {
+	return &models.Correction{
+		Msg: msg,
+		F: func() error {
+			ctx := context.Background()
+			existingID := oldRec.Original.(DNSRecord).ID
+			return c.DeleteDNSRecord(ctx, domain, existingID)
+		},
+	}
+}
+
+// toVercelCreateRequest converts a RecordConfig to a Vercel CreateDNSRecordRequest.
+func toVercelCreateRequest(domain string, rc *models.RecordConfig) (createDNSRecordRequest, error) {
+	req := createDNSRecordRequest{}
+
+	req.Domain = domain
+
+	name := rc.GetLabel()
+	if name == "@" {
+		name = ""
+	}
+	req.Name = name
+	req.Type = rc.Type
+	req.Value = rc.GetTargetField()
+	req.TTL = int64(rc.TTL)
+	req.Comment = ""
+
+	switch rc.Type {
+	case "MX":
+		req.MXPriority = int64(rc.MxPreference)
+	case "SRV":
+		req.SRV = &vercelClient.SRV{
+			Priority: int64(rc.SrvPriority),
+			Weight:   int64(rc.SrvWeight),
+			Port:     int64(rc.SrvPort),
+			Target:   rc.GetTargetField(),
+		}
+		req.Value = "" // SRV uses the SRV struct, not Value
+	case "TXT":
+		req.Value = rc.GetTargetTXTJoined()
+	case "HTTPS":
+		req.HTTPS = &httpsRecord{
+			Priority: int64(rc.SvcPriority),
+			Target:   rc.GetTargetField(),
+			Params:   rc.SvcParams,
+		}
+	case "CAA":
+		req.Value = fmt.Sprintf(`%v %s "%s"`, rc.CaaFlag, rc.CaaTag, rc.GetTargetField())
+	}
+
+	return req, nil
+}
+
+// toVercelUpdateRequest converts a RecordConfig to a Vercel UpdateDNSRecordRequest.
+func toVercelUpdateRequest(rc *models.RecordConfig) (updateDNSRecordRequest, error) {
+	req := updateDNSRecordRequest{}
+
+	name := rc.GetLabel()
+	if name == "@" {
+		name = ""
+	}
+	req.Name = &name
+
+	value := rc.GetTargetField()
+	req.Value = &value
+
+	req.TTL = ptrInt64(int64(rc.TTL))
+	req.Comment = ""
+
+	switch rc.Type {
+	case "MX":
+		req.MXPriority = ptrInt64(int64(rc.MxPreference))
+	case "SRV":
+		req.SRV = &vercelClient.SRVUpdate{
+			Priority: ptrInt64(int64(rc.SrvPriority)),
+			Weight:   ptrInt64(int64(rc.SrvWeight)),
+			Port:     ptrInt64(int64(rc.SrvPort)),
+			Target:   &value,
+		}
+		req.Value = nil // SRV uses the SRV struct, not Value
+	case "TXT":
+		txtValue := rc.GetTargetTXTJoined()
+		req.Value = &txtValue
+	case "HTTPS":
+		req.HTTPS = &httpsRecord{
+			Priority: int64(rc.SvcPriority),
+			Target:   rc.GetTargetField(),
+			Params:   rc.SvcParams,
+		}
+	case "CAA":
+		value := fmt.Sprintf(`%v %s "%s"`, rc.CaaFlag, rc.CaaTag, rc.GetTargetField())
+		req.Value = &value
+	}
+
+	return req, nil
+}
+
+// ptrInt64 returns a pointer to an int64
```

Comment:
```suggestion
	req.TTL = lo.ToPtr(int64(rc.TTL))
	req.Comment = ""

	switch rc.Type {
	case "MX":
		req.MXPriority = ptrInt64(int64(rc.MxPreference))
	case "SRV":
		req.SRV = &vercelClient.SRVUpdate{
			Priority: ptrInt64(int64(rc.SrvPriority)),
			Weight:   ptrInt64(int64(rc.SrvWeight)),
			Port:     ptrInt64(int64(rc.SrvPort)),
			Target:   &value,
		}
		req.Value = nil // SRV uses the SRV struct, not Value
	case "TXT":
		txtValue := rc.GetTargetTXTJoined()
		req.Value = &txtValue
	case "HTTPS":
		req.HTTPS = &httpsRecord{
			Priority: int64(rc.SvcPriority),
			Target:   rc.GetTargetField(),
			Params:   rc.SvcParams,
		}
	case "CAA":
		value := fmt.Sprintf(`%v %s "%s"`, rc.CaaFlag, rc.CaaTag, rc.GetTargetField())
		req.Value = &value
	}

	return req, nil
}
```

If acceptable by dnscontrol maintainers, github.com/samber/lo is recomended here for this helper with Generics.